### ISSUE TITLE
Soft loss on edges

### ISSUE THREAD

**Issue by Luugaaa** [ivadomed/model_seg_sc-gm-lesion_human_ms_exvivo_t2star#31]
# Context 

It can be challenging to accurately manually segment, especially on the edges of objects (whether it's the WM, GM or the lesions). For this reason we decided to investigate the use of a soft loss on the edges.

# Method 

To compute the edges we use the gt masks. We create a dilated version and an eroded version of an object, the subtraction of the two gives us the edge with a certain thickness. We can then apply continuous weights to the pixels in the edges, those weights gradually increase when the pixel is closer to the border of the edge until they reach 1. 

For each class, we qualitatively chose the following values based on the typical size of an object (if the edges are too thick we could allow for small objects to disappear) and the overall confidence we had when annotating the data. The edge weight will be the lowest weight, in the middle of the edge. 

```python
        self.edge_params = {
            0: {'edge_weight': 0.9, 'kernel_size': 7},
            1: {'edge_weight': 0.9, 'kernel_size': 7},
            2: {'edge_weight': 0.8, 'kernel_size': 5},  
            3: {'edge_weight': 0.7, 'kernel_size': 5},
            4: {'edge_weight': 0.5, 'kernel_size': 5}
        }
```

## Examples

<img width="397" height="422" alt="Image" src="https://github.com/user-attachments/assets/b255abb9-291d-4847-9abc-1239934b3902" />

<img width="397" height="422" alt="Image" src="https://github.com/user-attachments/assets/e4d91cd0-2928-4f92-8359-90c1af4e79a4" />

# Results

Note: The following results are the average dice score over 5 folds on the validation set of the "full" 12 subjects dataset.

The colors are: 
- **royal blue:** Models trained with a soft loss on the edges
- **red:** Base models
- 
## Foreground Dice Score 

<img width="730" height="422" alt="Image" src="https://github.com/user-attachments/assets/7103e74d-6e07-4bfa-a553-37333b62b885" />

## Losses

**Validation Loss (Total)**
<img width="730" height="422" alt="Image" src="https://github.com/user-attachments/assets/8c198a3a-3439-4417-b630-95f071ff829c" />

**Validation Loss (Dice)**

<img width="730" height="422" alt="Image" src="https://github.com/user-attachments/assets/0b345f69-ab62-440f-a683-f5a2dae69d84" />

**Validation Loss (CE)**

<img width="730" height="422" alt="Image" src="https://github.com/user-attachments/assets/32862089-311e-4f9a-ba95-d4113bca84de" />

**Training Loss (total)**

<img width="730" height="422" alt="Image" src="https://github.com/user-attachments/assets/ae24f0b6-4dbd-4780-8a42-d751cd1d3f89" />

## Observation
The models trained with the soft loss show the same dice results as the ones trained with a classic loss. 
However, we note that the models trained with soft loss overfit less than the base ones on the CE loss (only). That's an interesting behavior. 

# Conclusion

We don't observe an obvious improvement in the segmentation (identical validation dice scores), but we notice an attenuation of the overfitting when using the soft edge loss. Therefore, we believe we should conduct a qualitative evaluation before we can conclude anything.

**Comment by Luugaaa** [comment]
# Quantitative Analysis

## Subject TNU025, chunk S2

### Longitudinal 

**Base Model:**

<img width="822" height="298" alt="Image" src="https://github.com/user-attachments/assets/656ffe61-872d-4c9d-8f64-c93e397cfa68" />

**Soft Loss Model:**

<img width="822" height="298" alt="Image" src="https://github.com/user-attachments/assets/b0c4a54a-dbbd-46ac-ad27-205db60388a7" />

### Coronal Section 

**Base Model:**

<img width="815" height="282" alt="Image" src="https://github.com/user-attachments/assets/a89bc24b-d09a-4c2a-99d8-52d98be1bded" />

**Soft Loss Model:**

<img width="815" height="282" alt="Image" src="https://github.com/user-attachments/assets/f80b779f-4ac2-4816-82bf-416806ffcb7f" />

**Ref:**

<img width="815" height="282" alt="Image" src="https://github.com/user-attachments/assets/872330cd-9b01-4860-a106-81c6126dd3f8" />

### Observation

| Object | Base Model | Soft Loss |
| :------- | :----------: | --------: |
| Material on the left | NG | OK |
| Gray Matter | OK | NG |
| Lesion in Gray Matter | NG | OK |
